c2pa_c_ffi/src/c_api.rs

```diff
@@ -17,6 +17,93 @@ use std::{
     ptr,
 };
 
+/// Validates that a buffer size is within safe bounds and doesn't cause integer overflow
+/// when used with pointer arithmetic.
+///
+/// # Arguments
+/// * `size` - Size to validate
+/// * `ptr` - Pointer to validate against (for address space checks)
+///
+/// # Returns
+/// * `true` if the size is safe to use
+/// * `false` if the size would cause integer overflow
+unsafe fn is_safe_buffer_size(size: usize, ptr: *const c_uchar) -> bool {
+    // Combined checks for early return - improves branch prediction
+    if size == 0 || size > isize::MAX as usize {
+        return false;
+    }
+
+    // Check if the buffer would extend beyond address space to fail fast
+    if !ptr.is_null() {
+        let end_ptr = ptr.add(size);
+        if end_ptr < ptr {
+            return false; // Wrapped around
+        }
+    }
+
+    true
+}
+
+/// Creates a safe slice from raw parts with bounds validation
+///
+/// # Arguments
+/// * `ptr` - Pointer to the data
+/// * `len` - Length of the data
+/// * `param_name` - Name of the parameter for error reporting
+///
+/// # Returns
+/// * `Ok(slice)` if the slice is safe to create
+/// * `Err(Error)` if bounds validation fails
+unsafe fn safe_slice_from_raw_parts(
+    ptr: *const c_uchar,
+    len: usize,
+    param_name: &str,
+) -> Result<&[u8], Error> {
+    if ptr.is_null() {
+        return Err(Error::NullParameter(param_name.to_string()));
+    }
+
+    if !is_safe_buffer_size(len, ptr) {
+        return Err(Error::Other(format!(
+            "Buffer size {} is invalid for parameter '{}'",
+            len, param_name
+        )));
+    }
+
+    Ok(std::slice::from_raw_parts(ptr, len))
+}
+
+/// Validates a format string, which is either a mimetype,
+/// or an extension (eg `jpg`).
+///
+/// # Arguments
+/// * `format` - The format string to validate (mimetype or extension)
+///
+/// # Returns
+/// * `Ok(())` if the format string is valid
+/// * `Err(Error)` if the format string is invalid
+fn validate_format_string(format: &str) -> Result<(), Error> {
+    for &byte in format.as_bytes() {
+        if byte == 0 {
+            return Err(Error::Other(
+                "Format string contains null bytes".to_string(),
```

Comment:
can from_cstr return a string with null bytes? if it can, then we may need this check there, if not, we don't need it here.

Reply:
I managed to have strings containing only null bytes, albeit in convoluted test cases.